### ISSUE TITLE
Intermittent `RedisClient::ReadTimeoutError` and `RedisClient::CannotConnectError` issues.

### ISSUE THREAD

**Issue by MrJoy** [sidekiq/sidekiq#5929]
Ruby version: 3.1.4
Rails version: 7.0.4.2
Sidekiq / __Pro__ / Enterprise version(s): 7.1.1

I apologize for the lack of greater diagnostic information than is provided here.  Right now, I'm mostly looking for either "oh, you're doing X obviously wrong", or "here's what you should look at".

We're running a single Sidekiq node on a `t3.xlarge`, talking to an ElastiCache Redis instance of type `cache.t3.small` which is dedicated to Sidekiq.  The only other things talking to it are the Rails node (to fire jobs), and a standalone Ruby server process (also to fire jobs).

This is happening in our production environment, but not our staging environment.  The production environment is handling a relatively small job load, but it's 10x greater than our staging environment (although staging has a smaller Sidekiq instance type).  Production is handling about 500,000 jobs per day at the moment.

On the Redis server:

CPU usage is never going about bout 2%, and the spikes that high do not correlate to when the timeouts happen.  Engine CPU utilization peaks at around 0.5% and also doesn't correlate.

Network bytes in peaks at under 1.5M but again, doesn't correlate.  Network bytes out has similar peaks and also doesn't correlate.  Packets in peaks around 7k, packets out at 8.3k.  Again, no correlation with timeouts.  None of the "allowance exceeded" metrics in CloudWatch are going above 0.

On the EC2 instance:

CPU usage peaks around 11%, bytes in peaks around 8.1M, bytes out at 6.5M, packets in/out both at 21.5k -- none of those metrics' spikes correlating with the timeouts.

The parameter group for the Redis instance is the `redis7` config with the only change being that `maxmemory-policy` is explicitly set to `noeviction`.

Initializer/sidekiq config/stack traces here: https://gist.github.com/MrJoy/38b15c98ce58aef401f867a54589d9cc

Relevant config values:

* `Rails.configuration.sidekiq[:reliable_push]`: `true`
* `Rails.configuration.sidekiq[:max_conns]`: `32`
* `Rails.configuration.sidekiq[:critical_concurrency]`: `8`
* `Rails.configuration.sidekiq[:low_concurrency]`: `8`
* `Rails.configuration.sidekiq[:heavy_concurrency]`: `8`
* `ENV.fetch("SIDEKIQ_CONCURRENCY")`: `8`

There are 4 Sidekiq processes running on the server, for a total of 128 threads.

I'm not sure what exactly changed here.  We've had this degree of load since 4/26, but the errors only started showing up on 5/21.

<img width="1225" alt="image" src="https://github.com/sidekiq/sidekiq/assets/113625/aedfd9bd-9189-41ff-8799-cac89adfc731">

I'm at a loss for what to look for that might help me diagnose the problem.  How do you recommend I proceed?

**Comment by mperham** [comment]
I have no idea. Nothing immediately rings a bell or seems suspicious. Your concurrency is high IMO, 32 threads per process is likely to oversaturate a CPU and potentially lead to timeouts. Do you see any warnings logged from Sidekiq?

https://github.com/sidekiq/sidekiq/blob/0175cdbe03c3a9edf908a155ae6f006061b5c224/lib/sidekiq/launcher.rb#L213-L219

**Comment by MrJoy** [comment]
Our workload is _supremely_ I/O bound.  Every single job we have spends an overwhelming majority of its time on blocking I/O.  Thus the incredibly low CPU usage on the Sidekiq boxes.

At some point we'll switch the brunt of our jobs into Go and use Faktory, but I'm hoping to postpone that until at least after our big public launch.

**Comment by MrJoy** [comment]
BTW, we _are_ using Dogstatsd with Sidekiq Pro.  Any chance of getting RTT readings recorded as a metric?  I didn't see it in the metrics explorer, so I assume it's not being recorded right now.

**Comment by mperham** [comment]
It sounds like you're a lot more savvy than the average Sidekiq user, apologies if my response was simplistic. I can't tell how well a given user knows Sidekiq. Your concurrency sounds well tuned.

Unfortunately RTT is calculated in Sidekiq OSS which does not have Statsd support. Sidekiq Pro only collects metrics per-job so I'd have to collect RTT with every job execution, which is overkill. I could collect it in Sidekiq Enterprise's historical metrics.

We have other users seeing occasional mysterious Redis errors, like [here](https://github.com/redis-rb/redis-client/issues/119). That issue sounds different from yours but you might skim through the closed issues and see if anything rings a bell.

**Comment by MrJoy** [comment]
I'll take a look, thanks.  Totally understand on the metrics side of thing.  Unfortunately, my org isn't in a position where Enterprise is a prudent investment just yet but I look forward to the day we are.

I've done the work of consolidating the bulk of those periodic jobs, and am also considering the idea of doubling the number of processes and halving the number of threads just in case it's a GVL-related issue.  I can also double the process count / half the thread count if that does turn out to be the case, since we have plenty of RAM headroom.

This seems to be a very sporadic problem right now, cropping up 2-3 times per day, so it'll be hard to achieve any certainty about the efficacy of any changes.  Will update with any notes as they come up.

**Comment by mperham** [comment]
More processes, less threads is always more flexible IMO. Would AWS support be able to help you with ElastiCache issues? Maybe they can correlate the storm of errors with some other incident.

**Comment by MrJoy** [comment]
We have a meet-and-greet with our newly-assigned AWS Account Rep coming up soon.  Since their role largely seems to be "push on tickets to make sure they get addressed", filing a support ticket may be a practical option soon.  That said, we are smol, and aren't paying for support yet so my expectations are low.

In the meantime, I'm gonna try these two changes (one at a time, ofc) and go spelunking in the closed issues for more insight.

**Comment by MrJoy** [comment]
Just for the sake of capturing data (not sure if it's remotely relevant or not):

The errors seem to have gone from more-sporadic-but-more-common-when-they-happen to more-regular-but-less-frequent when we changed from the `redis-client` pure-Ruby driver to hiredis:

* `ReadTimeoutError`, pre-Hiredis:
    * 17th: 7
    * 18th: 1
    * 21st: 4
* `ReadTimeoutError`, post-Hiredis:
    * 21st: 2
    * 22nd: 2
    * 23rd: 3

Our deploys to production recently have been on (in UTC): The 2nd, the 17th, and the 21st.  On the 21st, the only change was to move to hiredis.

The first error occurred at `2023-05-17 19:08:46 UTC`.  That is after our deploy on the 17th, which was around `2023-05-17 17:44 UTC`.  So, that _is_ strongly correlated.

Looking through the diffs, there are no substantive changes to Sidekiq jobs or the code that they depend on:

* I added a little instrumentation to our wrapper around the client library for the Google API (adding a tag to the Datadog metric that gets recorded on each call), and touched all the sites where an instance of that wrapper is created in order to pass the extra information through.
* We renamed two Sidekiq jobs.  (Neither of which runs on a schedule that correlates with the errors.)
* We jettisoned two unused columns from a table.
* In the event of a highly uncommon and very specific error from Google, we handle things slightly differently involving an extra DB write.

All other code changes were in code that can't be called from Sidekiq jobs (e.g. our GraphQL API, and our admin UI).

We did, however, update some dependencies.

The ones that have any potential relevance (I.E. are in the Sidekiq jobs' codepaths, and have any sort of network/threading involvement) are:

* `sidekiq-pro`: 7.0.9 => 7.1.1
* `sidekiq`: 7.0.8 => 7.1.0
* `datadog_api_client`: 2.9.0 => 2.10.0
* `google-apis-calendar_v3`: 0.27.0 => 0.28.0
* `google-apis-gmail_v1`: 0.25.0 => 0.27.0
* `googleauth`: 1.5.1 => 1.5.2
* `parallel`: 1.22.1 => 1.23.0
* `pg`: 1.4.6 => 1.5.3
* `puma`: 6.2.1 => 6.2.2
* `rack`: 2.2.6.4 => 2.2.7
* `sentry-rails`: 5.8.0 => 5.9.0
* `sentry-ruby`: 5.8.0 => 5.9.0

I am inclined to download the lot of them to see if that makes the issue go away.  Will provide updates with anything I discover.

**Comment by MrJoy** [comment]
So far, no more errors. Of course, that's not dispositive since (at least before putting hiredis in place, which I've removed for this test) the errors weren't happening _every_ day.  I'm going to wait another day with the current configuration (modulo unrelated, out-of-codepath changes that need to go out).  If I don't see anything by CoB tomorrow, I'll re-add hiredis since that seemed to make things happen a little more consistently.  If I still don't see anything another day out, then I'll update the gems and see if the issue returns.  If it does, then I'll track down exactly which gem update is relevant.  If not, then...  Heisenbug, FTL.

**Comment by MrJoy** [comment]
And my cunning plan to provoke the universe into throwing errors at my by writing an "all clear so far" update worked.  Just got a flurry of the errors.  The gem updates are unrelated to the issue.

**Comment by MrJoy** [comment]
Interestingly, all four running Sidekiq processes encountered the error within ~200ms in the most recent burst.  So this seems to be an issue that crosses processes.  That would seem to make issues like long GC pauses less likely as a culprit.

<img width="376" alt="image" src="https://github.com/sidekiq/sidekiq/assets/113625/67bf2ebc-3680-4a97-9383-14e40c9c8906">

**Comment by mperham** [comment]
Is it possible this is an issue on the remote Redis side? Does ElastiCache migrate or move instance locations "live" or do you opt into a migration window? What Redis version are you using?

**Comment by mperham** [comment]
Sounds like you are progressing down a reasonable path of debugging. Keep it up!

**Comment by gavin-itison** [comment]
By way of a +1, we're seeing similar issues since updating to: `sidekiq-ent 7.1.0` on `ruby 3.2.2` with `rails 6.1.7.1`.

We've not started a deep dive on debugging yet, but I can say we're not on AWS, are talking to vanilla Redis `6.2.12` and the server doesn't appear to be particularly taxed on any of our high-level monitoring when the outages occur.

**Comment by MrJoy** [comment]
@gavin-itison We were on Sidekiq 7.0.8, and Sidekiq Pro 7.0.9, but as I noted, rolling back to that version did not resolve the issue.

**Comment by MrJoy** [comment]
After leaving two nodes running for a few days, I saw no errors.  Spinning back down to one node (and also deploying pending changes out of necessity), I have seen one occurrence in the past 24 hours.

Still monitoring to see if I see clusters going forward, but this makes me feel that the issue is less network / Redis, and more my configuration / workload.  I'm going to gather a bit more data, and then depending on what I see I may try cutting the worker count from 32 to 24.

**Comment by nateberkopec** [comment]
That is a lot of load you're pushing through some very small AWS instances.

If I had to guess, I would think that you're hitting some hardware limitations. t3.smalls are 2 vCPU shared instances. It's entirely possible that 128 Sidekiq threads (particularly when they _all_ start up at once at a deploy) could overwhelm that rather puny amount of CPU capacity available, causing a handful of timeouts like you're seeing.

I don't believe your utilization metrics because it only takes a 5-second long spike in CPU (again, as in 128 Sidekiq threads all trying to do the same thing at the same time) to cause this issue and that may get averaged out in a longer-term CPU utilization measurement.

**Comment by nateberkopec** [comment]
> At some point we'll switch some of the jobs to using async I/O and bundling many of them together (possibly also taking the opportunity to move our core jobs into Go and Faktory), but I'm rather hoping to avoid that until at least January.

Probably the most cost-effective thing you could try would be to migrate that Redis instance to something a lot bigger... I'd try m6g.xlarge just to see if it solved the problem entirely.

**Comment by MrJoy** [comment]
> If I had to guess, I would think that you're hitting some hardware limitations. t3.smalls are 2 vCPU shared instances. It's entirely possible that 128 Sidekiq threads (particularly when they _all_ start up at once at a deploy) could overwhelm that rather puny amount of CPU capacity available, causing a handful of timeouts like you're seeing.

One thing to note is that the 128 threads is spread across 4 vCPUs (`t3.xlarge`), with CPU-pinning per-process (32 threads per process).

It doesn't feel like the Redis instance is getting overwhelmed here, because the problem went away when I doubled the number of Sidekiq worker nodes (total of 256 threads).  That strikes me as suggesting that it's an issue on the SideKiq worker node itself.

CPU usage is certainly the #1 candidate here, and yeah, the metrics could be too coarse-grained to capture spikes, but the thing that has my spidey-sense tingling there is that I've seen a few occasions where the error comes from _all four processes_ within the span of a couple hundred milliseconds.  Given that processes are pinned to CPUs, it seems quite the coincidence that all four workers would happen to jam up all four CPUs at once.  

I'm having difficulty _forcing_ the error by scheduling a ton of jobs, which is making it harder for me to pin things down more precisely.  I may try _increasing_ the worker count to try and induce more frequent errors.  If it's really a CPU load issue, that should help confirm that.  If increasing the thread count amplifies the problem, I'll try reducing it to 24 and see if that makes things go away.

**Comment by krasnoukhov** [comment]
Not sure if it's going to be helpful, but we've seen similar exceptions pop up after upgrade to Sidekiq 7 specifically. We use batches that push up to 500k jobs, and it seems like these exceptions are kind of tied to that process. If that's the case for you (meaning that this started happening after upgrade) I've noted one thing that changed slightly unnoticed between Sidekiq 6 and 7.

Before, it seems like redis timeout was set to [5 seconds by default](https://github.com/redis/redis-rb/blob/ff4399bfbc9d458bc1de1634a9714ad6767348dd/lib/redis.rb#L48) and since the switch to redis-client it's now [1 second by default](https://github.com/redis-rb/redis-client/blob/7f76e30799b53d5d84bc1b8907f435840ebde468/README.md?plain=1#L79). I've configured our sidekiq to have 5 second timeout and it seems like it helped so far...

**Comment by krasnoukhov** [comment]
Good point @mperham, totally makes sense to keep it all speedy. I personally think current default is alright, given that it's actually [documented](https://github.com/sidekiq/sidekiq/wiki/Using-Redis#life-in-the-cloud) in the wiki 😅 

Not sure if this issue is a good place to ask, but basically the problem for us is large batch jobs. We use batch feature to push up to 500k jobs in transaction-like manner so if something goes wrong in between parent job fails, nothing is queued, and it retries. Aside from redesigning the job logic so it does not use batch (allows idempotency/duplicates), any tips out there? Maybe batch push can be improved in a way that it breaks down job payloads into smaller pieces to not lock redis too much?

**Comment by gavin-itison** [comment]
In our case it isn't a big batch that's affected, but a worker with a `concurrent` `wait_timeout` of 5 seconds

I've not went code-spelunking yet to see how `wait_timeout` is implemented (so apologies if this is the wrong track), but in older version of sidekiq I think I'm right in saying the default `wait_timeout` of 5s happened to match with the timeout on the connection. I was planning on looking at whether a `wait_timeout > redis timeout` was problematic.

**Comment by mperham** [comment]
> Not sure if this issue is a good place to ask, but basically the problem for us is large batch jobs. We use batch feature to push up to 500k jobs in transaction-like manner so if something goes wrong in between parent job fails, nothing is queued, and it retries. Aside from redesigning the job logic so it does not use batch (allows idempotency/duplicates), any tips out there? Maybe batch push can be improved in a way that it breaks down job payloads into smaller pieces to not lock redis too much?

It's possible. The idea would be to push batch jobs into a thread-specific staging area in Redis, N at a time, and then have a "commit" phase which makes the staging area "live". I have no clue how that might be implemented inside `Sidekiq::Batch` or `Sidekiq::Client` but it sounds like some fun Sidekiq R&D if anyone else wants to hack on it. I won't be able to get to this until Aug/Sep.

**Comment by renanigt** [comment]
> In our case it isn't a big batch that's affected, but a worker with a `concurrent` `wait_timeout` of 5 seconds
> 
> I've not went code-spelunking yet to see how `wait_timeout` is implemented (so apologies if this is the wrong track), but in older version of sidekiq I think I'm right in saying the default `wait_timeout` of 5s happened to match with the timeout on the connection. I was planning on looking at whether a `wait_timeout > redis timeout` was problematic.

I faced this same issue with a `concurrent` limiter. It worked well when using sidekiq 6 then when we upgraded it to the latest version se started getting this error:
```
Worker' just died with error RedisClient::ReadTimeoutError: "Waited 1.0002023898996413 seconds".
```

**Comment by andrewsheelan** [comment]
@renanigt - so far we haven't seen the issue after we rolled back to 7.1.2
Here are the versions we have now:
sidekiq (7.1.2) 
sidekiq-pro (7.1.3)
sidekiq-ent (7.1.1)

We are testing in a different environment with the latest versions and sidekiq 7.1.6. will keep you posted.

**Comment by renanigt** [comment]
@mperham 

do you see any reason for upgrading to sidekiq (7.1.6) would sort out the issue with timeouts when using `concurrent` limiter?

**Comment by mperham** [comment]
@renanigt I don't understand.

**Comment by renanigt** [comment]
hey @mperham 

after upgrading sidekiq to 7.1.2, I ran into a situation in which a worker had a concurrent limiter with `wait_timeout: 3` started raising errors like this: "Worker' just died with error RedisClient::ReadTimeoutError: "Waited 1.0002023898996413 seconds."

Similar what @gavin-itison and @andrewsheelan mentioned above as well:
>In our case it isn't a big batch that's affected, but a worker with a concurrent wait_timeout of 5 seconds
>
>I've not went code-spelunking yet to see how wait_timeout is implemented (so apologies if this is the wrong track), but in older version of sidekiq I think I'm right in saying the default wait_timeout of 5s happened to match with the timeout on the connection. I was planning on looking at whether a wait_timeout > redis timeout was problematic.

when we change it to 0 it works well... is there any reason/situation where a wait_timeout > redis timeout would raise a RedisClient::ReadTimeoutError?

@andrewsheelan mentioned that upgrading to `7.1.6` made the error not happen anymore.

Just curious if there's any reason for this.

**Comment by mperham** [comment]
I don't think I ever understood the root cause here. It's possible a combo of changes in redis-client 0.x and in Sidekiq proper fixed the issue.

**Comment by ksweetie** [comment]
@mperham We're also seeing these errors when using a Concurrent Limiter using 7.1.1. [This comment](https://github.com/sidekiq/sidekiq/issues/5857#issuecomment-1645486651) on a separate issue describes our problem. A sample error:

```
RedisClient::ReadTimeoutError: Waited 1.0000936759752221 seconds
```

Our first thought was to increase the `read_timeout` to 2, but we still get these `Waited 1.0000X... seconds` errors. I was confused why it was only waiting just over 1 second when the read timeout was supposed to be 2 seconds. That appears to be coming from the hardcoded `diff + 1` shown in the linked comment.

Do you have a recommendation for how to fix?

**Comment by mperham** [comment]
@ksweetie What happens when you upgrade to a version where people are reporting this issue fixed, e.g. 7.1.6?

**Comment by ksweetie** [comment]
@mperham It's a bit confusing that there are two GH issues here that I think are talking about the same error. In [this issue](https://github.com/sidekiq/sidekiq/issues/5857#issuecomment-1685021916), it sounded like you acknowledge the cause, but are unsure of the right solution. In this issue you're saying you don't understand the root cause. AFAICT, both issues are talking about the same problem in the concurrent limiter, but if these are distinct issues I can move my comments to the other one. The relevant code in 7.1.2(ent) looks unchanged:

```rb
candidate = conn.blocking_call(diff + 1, "BLMOVE", @free, @pend, "RIGHT", "LEFT", diff.to_f)
```

The hardcoded `diff + 1` can result in errors that look like:

```
RedisClient::ReadTimeoutError: Waited 1.0000936759752221 seconds
```

even with a `read_timeout` of 2. Shouldn't this take `read_timeout` into account? I'm not sure how to get around the hardcoded 1 here.

**Comment by mperham** [comment]
@ksweetie I'm not experiencing the problem myself so I can only pass along what others said. My confusion stems from this user report which says the problem seems to be gone.

> Quick update, sidekiq (7.1.6), sidekiq-pro (7.1.6), sidekiq-ent (7.1.2) - looks good so far in our tests. We are going ahead with these versions.

Are you running those versions and still seeing the issue?

**Comment by andrewsheelan** [comment]
Sorry guys, its still happening with the latest specified versions, we started seeing them again this week with bigger loads. The fails don't directly place them inside the dead queue anymore (so retries and finishes the batch) but it does seem to happen still. Not sure if it helps but these are our redis gem versions - redis (5.0.8), redis-client (0.17.1)

**Comment by MrJoy** [comment]
> @ksweetie What happens when you upgrade to a version where people are reporting this issue fixed, e.g. 7.1.6/7.1.6/7.1.2?

@mperham Just as a heads-up:  We've seen this error happen at least once since upgrading to 7.1.6.  Possibly a bit less frequently than before, but it's not occurring frequently enough for me to be confident of that.

**Comment by mperham** [comment]
I'm struggling to reproduce this; I don't understand how this is happening. Can someone give me a full backtrace with the latest versions?

**Comment by renanigt** [comment]
hey @mperham, this is what I have:
```console
RedisClient::ReadTimeoutError: Waited 1.0003620809875429 seconds
block in fill_buffer(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/ruby_connection/buffered_io.rb:140)
loop(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/ruby_connection/buffered_io.rb:122)
fill_buffer(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/ruby_connection/buffered_io.rb:122)
ensure_remaining(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/ruby_connection/buffered_io.rb:114)
getbyte(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/ruby_connection/buffered_io.rb:85)
parse(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/ruby_connection/resp3.rb:113)
load(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/ruby_connection/resp3.rb:50)
block in read(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/ruby_connection.rb:96)
with_timeout(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/ruby_connection/buffered_io.rb:48)
read(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/ruby_connection.rb:96)
call(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/connection_mixin.rb:31)
block (2 levels) in blocking_call(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client.rb:285)
call(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/middlewares.rb:16)
block in blocking_call(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client.rb:284)
ensure_connected(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client.rb:643)
blocking_call(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client.rb:283)
blocking_call(/app/vendor/bundle/ruby/3.2.0/gems/redis-client-0.17.0/lib/redis_client/decorator.rb:26)
block in lock(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-ent-7.1.2/lib/sidekiq-ent/limiter/concurrent.rb:109)
block in redis(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-ent-7.1.2/lib/sidekiq-ent/limiter.rb:103)
block (2 levels) in with(/app/vendor/bundle/ruby/3.2.0/gems/connection_pool-2.4.1/lib/connection_pool.rb:110)
handle_interrupt(/app/vendor/bundle/ruby/3.2.0/gems/connection_pool-2.4.1/lib/connection_pool.rb:109)
block in with(/app/vendor/bundle/ruby/3.2.0/gems/connection_pool-2.4.1/lib/connection_pool.rb:109)
handle_interrupt(/app/vendor/bundle/ruby/3.2.0/gems/connection_pool-2.4.1/lib/connection_pool.rb:106)
with(/app/vendor/bundle/ruby/3.2.0/gems/connection_pool-2.4.1/lib/connection_pool.rb:106)
redis(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-ent-7.1.2/lib/sidekiq-ent/limiter.rb:100)
lock(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-ent-7.1.2/lib/sidekiq-ent/limiter/concurrent.rb:104)
within_limit(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-ent-7.1.2/lib/sidekiq-ent/limiter/concurrent.rb:61)
perform_within_limit(/app/app/workers/yardi/base_worker.rb:12)
perform(/app/app/workers/yardi/invite_resident_worker.rb:9)
execute_job(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:210)
block (4 levels) in process(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:180)
traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:180)
block in traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:183)
call(/app/lib/sidekiq/middleware/server/args_saver.rb:4)
traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:182)
block in traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:183)
call(/app/vendor/bundle/ruby/3.2.0/gems/rollbar-3.4.0/lib/rollbar/plugins/sidekiq/plugin.rb:11)
traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:182)
block in traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:183)
block in call(/app/vendor/bundle/ruby/3.2.0/gems/newrelic_rpm-8.15.0/lib/new_relic/agent/instrumentation/sidekiq/server.rb:25)
perform_action_with_newrelic_trace(/app/vendor/bundle/ruby/3.2.0/gems/newrelic_rpm-8.15.0/lib/new_relic/agent/instrumentation/controller_instrumentation.rb:383)
call(/app/vendor/bundle/ruby/3.2.0/gems/newrelic_rpm-8.15.0/lib/new_relic/agent/instrumentation/sidekiq/server.rb:20)
traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:182)
block in traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:183)
call(/app/vendor/bundle/ruby/3.2.0/gems/marginalia-sidekiq-0.1.0/lib/marginalia/sidekiq.rb:8)
traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:182)
block in traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:183)
call(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-pro-7.1.4/lib/sidekiq/batch/middleware.rb:58)
traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:182)
block in traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:183)
call(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-ent-7.1.2/lib/sidekiq-ent/unique.rb:182)
traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:182)
block in traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:183)
call(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-ent-7.1.2/lib/sidekiq-ent/limiter/middleware.rb:10)
traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:182)
block in traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:183)
track(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/metrics/tracking.rb:26)
call(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/metrics/tracking.rb:122)
traverse(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:182)
invoke(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/middleware/chain.rb:173)
block (3 levels) in process(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:179)
block (6 levels) in dispatch(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:140)
local(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/job_retry.rb:114)
block (5 levels) in dispatch(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:139)
block in call(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/rails.rb:16)
wrap(/app/vendor/bundle/ruby/3.2.0/gems/activesupport-6.1.7.6/lib/active_support/execution_wrapper.rb:91)
block in wrap(/app/vendor/bundle/ruby/3.2.0/gems/activesupport-6.1.7.6/lib/active_support/reloader.rb:72)
wrap(/app/vendor/bundle/ruby/3.2.0/gems/activesupport-6.1.7.6/lib/active_support/execution_wrapper.rb:91)
wrap(/app/vendor/bundle/ruby/3.2.0/gems/activesupport-6.1.7.6/lib/active_support/reloader.rb:71)
call(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/rails.rb:15)
block (4 levels) in dispatch(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:135)
stats(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:271)
block (3 levels) in dispatch(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:130)
call(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/job_logger.rb:13)
block (2 levels) in dispatch(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:129)
global(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/job_retry.rb:81)
block in dispatch(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:128)
prepare(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/job_logger.rb:39)
dispatch(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:127)
block (2 levels) in process(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:178)
handle_interrupt(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:177)
block in process(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:177)
handle_interrupt(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:176)
process(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:176)
process_one(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:82)
run(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/processor.rb:72)
watchdog(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/component.rb:10)
block in safe_thread(/app/vendor/bundle/ruby/3.2.0/gems/sidekiq-7.1.4/lib/sidekiq/component.rb:19)
```

**Comment by mperham** [comment]
@renanigt Those aren't the latest versions. Can you do a `bundle up sidekiq-ent` and still reproduce it?

**Comment by renanigt** [comment]
Sorry @mperham, just noticed you had asked for "latest Sidekiq versions"... this error was from production because I couldn't reproduce it on my local as well, but I'll try to bundle it up and see what happens.

@MrJoy would you have the full backtrace?

**Comment by MrJoy** [comment]
From Sentry:

```
hiredis-client (0.18.0) lib/redis_client/hiredis_connection.rb in _read at line 95
hiredis-client (0.18.0) lib/redis_client/hiredis_connection.rb in read at line 95
redis-client (0.18.0) lib/redis_client/connection_mixin.rb in call at line 31
redis-client (0.18.0) lib/redis_client.rb in block (2 levels) in blocking_call at line 317
redis-client (0.18.0) lib/redis_client/middlewares.rb in call at line 16
redis-client (0.18.0) lib/redis_client.rb in block in blocking_call at line 316
redis-client (0.18.0) lib/redis_client.rb in ensure_connected at line 675
redis-client (0.18.0) lib/redis_client.rb in blocking_call at line 315
redis-client (0.18.0) lib/redis_client/decorator.rb in blocking_call at line 26
sidekiq-pro (7.1.6) lib/sidekiq/pro/super_fetch.rb in block in strict at line 356
sidekiq (7.1.6) lib/sidekiq/capsule.rb in block in redis at line 102
connection_pool (2.4.1) lib/connection_pool.rb in block (2 levels) in with at line 110
connection_pool (2.4.1) lib/connection_pool.rb in handle_interrupt at line 109
connection_pool (2.4.1) lib/connection_pool.rb in block in with at line 109
connection_pool (2.4.1) lib/connection_pool.rb in handle_interrupt at line 106
connection_pool (2.4.1) lib/connection_pool.rb in with at line 106
sidekiq (7.1.6) lib/sidekiq/capsule.rb in redis at line 99
sidekiq (7.1.6) lib/sidekiq/component.rb in redis at line 28
sidekiq-pro (7.1.6) lib/sidekiq/pro/super_fetch.rb in strict at line 342
sidekiq-pro (7.1.6) lib/sidekiq/pro/super_fetch.rb in call at line 313
sidekiq-pro (7.1.6) lib/sidekiq/pro/super_fetch.rb in get_job at line 313
sidekiq-pro (7.1.6) lib/sidekiq/pro/super_fetch.rb in block in retrieve_work at line 301
connection_pool (2.4.1) lib/connection_pool.rb in block (2 levels) in with at line 110
connection_pool (2.4.1) lib/connection_pool.rb in handle_interrupt at line 109
connection_pool (2.4.1) lib/connection_pool.rb in block in with at line 109
connection_pool (2.4.1) lib/connection_pool.rb in handle_interrupt at line 106
connection_pool (2.4.1) lib/connection_pool.rb in with at line 106
sidekiq-pro (7.1.6) lib/sidekiq/pro/super_fetch.rb in retrieve_work at line 300
sidekiq (7.1.6) lib/sidekiq/processor.rb in get_one at line 87
sidekiq (7.1.6) lib/sidekiq/processor.rb in fetch at line 99
sidekiq (7.1.6) lib/sidekiq/processor.rb in process_one at line 81
sidekiq (7.1.6) lib/sidekiq/processor.rb in run at line 72
sidekiq (7.1.6) lib/sidekiq/component.rb in watchdog at line 10
sidekiq (7.1.6) lib/sidekiq/component.rb in block in safe_thread at line 19
```

**Comment by le0pard** [comment]
sidekiq: 7.2.0
rails: 7.1.2
redis-client: 0.18.0

connection settings (timeout was increased, but not really help):

```ruby
  config.redis = {
    network_timeout: 3,
    pool_timeout: 3
  }
```

Redis instance in same VPC, as sidekiq. Using direct peering for traffic.

Trace:

```
RedisClient::ReadTimeoutError: Waited 5 seconds (RedisClient::ReadTimeoutError)
  from redis-client (0.18.0) lib/redis_client/ruby_connection/buffered_io.rb:140:in `block in fill_buffer'
  from redis-client (0.18.0) lib/redis_client/ruby_connection/buffered_io.rb:122:in `loop'
  from redis-client (0.18.0) lib/redis_client/ruby_connection/buffered_io.rb:122:in `fill_buffer'
  from redis-client (0.18.0) lib/redis_client/ruby_connection/buffered_io.rb:114:in `ensure_remaining'
  from redis-client (0.18.0) lib/redis_client/ruby_connection/buffered_io.rb:85:in `getbyte'
  from redis-client (0.18.0) lib/redis_client/ruby_connection/resp3.rb:113:in `parse'
  from redis-client (0.18.0) lib/redis_client/ruby_connection/resp3.rb:50:in `load'
  from redis-client (0.18.0) lib/redis_client/ruby_connection.rb:96:in `block in read'
  from redis-client (0.18.0) lib/redis_client/ruby_connection/buffered_io.rb:48:in `with_timeout'
  from redis-client (0.18.0) lib/redis_client/ruby_connection.rb:96:in `read'
  from redis-client (0.18.0) lib/redis_client/connection_mixin.rb:31:in `call'
  from redis-client (0.18.0) lib/redis_client.rb:317:in `block (2 levels) in blocking_call'
  from redis-client (0.18.0) lib/redis_client/middlewares.rb:16:in `call'
  from redis-client (0.18.0) lib/redis_client.rb:316:in `block in blocking_call'
  from redis-client (0.18.0) lib/redis_client.rb:675:in `ensure_connected'
  from redis-client (0.18.0) lib/redis_client.rb:315:in `blocking_call'
  from redis-client (0.18.0) lib/redis_client/decorator.rb:26:in `blocking_call'
  from sidekiq (7.2.0) lib/sidekiq/fetch.rb:47:in `block in retrieve_work'
  from sidekiq (7.2.0) lib/sidekiq/capsule.rb:102:in `block in redis'
  from connection_pool (2.4.1) lib/connection_pool.rb:110:in `block (2 levels) in with'
  from connection_pool (2.4.1) lib/connection_pool.rb:109:in `handle_interrupt'
  from connection_pool (2.4.1) lib/connection_pool.rb:109:in `block in with'
  from connection_pool (2.4.1) lib/connection_pool.rb:106:in `handle_interrupt'
  from connection_pool (2.4.1) lib/connection_pool.rb:106:in `with'
  from sidekiq (7.2.0) lib/sidekiq/capsule.rb:99:in `redis'
  from sidekiq (7.2.0) lib/sidekiq/component.rb:28:in `redis'
  from sidekiq (7.2.0) lib/sidekiq/fetch.rb:47:in `retrieve_work'
  from sidekiq (7.2.0) lib/sidekiq/processor.rb:87:in `get_one'
  from sidekiq (7.2.0) lib/sidekiq/processor.rb:99:in `fetch'
  from sidekiq (7.2.0) lib/sidekiq/processor.rb:81:in `process_one'
  from sidekiq (7.2.0) lib/sidekiq/processor.rb:72:in `run'
  from sidekiq (7.2.0) lib/sidekiq/component.rb:10:in `watchdog'
  from sidekiq (7.2.0) lib/sidekiq/component.rb:19:in `block in safe_thread'
```

In this time CPU and network (no load):

![image](https://github.com/sidekiq/sidekiq/assets/98444/446a5918-5722-4e20-975b-91562a646c4f)
![image](https://github.com/sidekiq/sidekiq/assets/98444/f6c81de2-a48f-4204-b0a9-9e481684088a)

In that time based on logs sidekiq doing nothing (no CPU intensive tasks):
![Screenshot 2023-11-24 at 23 45 25](https://github.com/sidekiq/sidekiq/assets/98444/8da63e47-635b-4f7e-bf3f-025a716209da)

BTW, sidekiq have concurrency 8 and only 6 reports in logs. So looks like 2 other threads dont have this issue:
![image](https://github.com/sidekiq/sidekiq/assets/98444/3f050c31-b56f-44cd-abf6-63f4bcd37032)
![Screenshot 2023-11-24 at 23 46 53](https://github.com/sidekiq/sidekiq/assets/98444/bf4d1938-b122-452e-b36a-e54ca7c0de6c)

**Comment by ksweetie** [comment]
@mperham Not sure if this will help, it looks similar to the ones above, but here's a trace with sidekiq-ent on the latest version. We're seeing anywhere from dozens to hundreds of these errors per week.

- Rails 7.0.7.1
- Sidekiq-ent 7.2.0
- redis-client 0.14.1

```
RedisClient::ReadTimeoutError: Waited 1.0000497290166095 seconds

[GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/ruby_connection/buffered_io.rb:140 :in `block in fill_buffer`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/ruby_connection/buffered_io.rb:122 :in `loop`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/ruby_connection/buffered_io.rb:122 :in `fill_buffer`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/ruby_connection/buffered_io.rb:114 :in `ensure_remaining`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/ruby_connection/buffered_io.rb:85 :in `getbyte`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/ruby_connection/resp3.rb:113 :in `parse`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/ruby_connection/resp3.rb:50 :in `load`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/ruby_connection.rb:96 :in `block in read`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/ruby_connection/buffered_io.rb:48 :in `with_timeout`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/ruby_connection.rb:96 :in `read`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/connection_mixin.rb:31 :in `call`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client.rb:272 :in `block (2 levels) in blocking_call`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/middlewares.rb:16 :in `call`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client.rb:271 :in `block in blocking_call`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client.rb:626 :in `ensure_connected`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client.rb:270 :in `blocking_call`
  [GEM_ROOT]/gems/redis-client-0.14.1/lib/redis_client/decorator.rb:26 :in `blocking_call`
 [GEM_ROOT]/gems/sidekiq-ent-7.2.0/lib/sidekiq-ent/limiter/concurrent.rb:109 :in `block in lock`
 [GEM_ROOT]/gems/sidekiq-ent-7.2.0/lib/sidekiq-ent/limiter.rb:103 :in `block in redis`
  [GEM_ROOT]/gems/connection_pool-2.3.0/lib/connection_pool.rb:65 :in `block (2 levels) in with`
  [GEM_ROOT]/gems/connection_pool-2.3.0/lib/connection_pool.rb:64 :in `handle_interrupt`
  [GEM_ROOT]/gems/connection_pool-2.3.0/lib/connection_pool.rb:64 :in `block in with`
  [GEM_ROOT]/gems/connection_pool-2.3.0/lib/connection_pool.rb:61 :in `handle_interrupt`
  [GEM_ROOT]/gems/connection_pool-2.3.0/lib/connection_pool.rb:61 :in `with`
 [GEM_ROOT]/gems/sidekiq-ent-7.2.0/lib/sidekiq-ent/limiter.rb:100 :in `redis`
 [GEM_ROOT]/gems/sidekiq-ent-7.2.0/lib/sidekiq-ent/limiter/concurrent.rb:104 :in `lock`
 [GEM_ROOT]/gems/sidekiq-ent-7.2.0/lib/sidekiq-ent/limiter/concurrent.rb:61 :in `within_limit`
```

**Comment by mperham** [comment]
@ksweetie What version of Redis server are you using? I've heard that OpenSSL 3 + some older versions of Redis can cause spurious connection errors.

**Comment by stanhu** [comment]
>  We kind of resolved the problem by increasing the Redis connection pool size in the Sidekiq initializer from the default value (10) to 50, like so

Could others try increasing the connection pool size like this? In Sidekiq 6, there was one connection pool of size `concurrency + 5`. In Sidekiq 7, a capsule has its own connection pool of size `concurrency`, and there is an internal Sidekiq connection pool with a fixed size of 10.

If you have Sidekiq jobs that run and access `Sidekiq.redis`--which accesses the capsule's connection pool--you might be running into an issue where a connection doesn't become available until a BRPOP returns from another job. It might be a good idea for Sidekiq 7 to give additional leeway by default. Until then, bumping up the pool is worth doing since connections are lazily created.

There's also a possibility that the 10 connections in the internal pool isn't sufficient, but most of the stack traces above appear to be coming from the job processors.

**Comment by MrJoy** [comment]
It's possible my instances of this are related to the connection count.  I failed to account for the fixed overhead.

I set Sidekiq's Redis pool size to the sum of the concurrency settings for non-capsule workers + all capsules.

I have not been adding 5 to that.  I will try adding that and see if this goes away.  Currently, we're getting little bursts every 4-7 days, so it may be a while before I can say whether that fixes it for us.

**Comment by ksweetie** [comment]
@mperham We're using Redis 6.2.13.

**Comment by mperham** [comment]
@ksweetie I'm not sure if that has the OpenSSL3 fix or not. Redis's release documentation is so copious I can't even find their version release notes anymore.

I would increase your pool size as Stan mentioned. If that doesn't work, consider an upgrade to Redis 7.x.

**Comment by MrJoy** [comment]
According to this, OpenSSL 3 fix is in 6.2.7: https://raw.githubusercontent.com/redis/redis/6.2/00-RELEASENOTES

**Comment by ksweetie** [comment]
We bumped the pool size yesterday, and so far we've seen one of these errors. The error is intermittent, so some days we have zero, other days hundreds, so it's hard to say if that helped some or not, but it looks like the issue is still present.

**Comment by MrJoy** [comment]
<img width="323" alt="image" src="https://github.com/sidekiq/sidekiq/assets/113625/31bebbbd-c186-4b94-b42b-092e2446032a">

I'll follow up in a couple more days, but so far it's looking like pool size was, in fact, the issue for us.

**Comment by MrJoy** [comment]
@mperham Perhaps an initialization check that queries the connection pool and ensures it's of sufficient size for the configured workers / capsules would help prevent further PEBKAC errors in the future?

**Comment by stanhu** [comment]
@MrJoy Do you have a good way to check that at boot time? It seems to me that will depend on the runtime behavior of the system. There should be some metric that tells you the P99 time it took to retrieve a connection from the pool because that number should be as close to 0 as possible.

For now, I think Sidekiq should default to adding 5 (to be consistent with Sidekiq 6) extra connections to the capsule pool so that there should be some extra connections that can be used by calls to `Sidekiq.redis` within a job that are not blocked by a BRPOP. This number should probably be configurable.

**Comment by MrJoy** [comment]
@stanhu I don't mean that Sidekiq should check the pool size dynamically, just that it's in the best position to do the math to determine if the requested pool size is adequate given the configured worker set size, capsule configurations, and internals.  Of course, it can only provide a check that the configured pool size is sufficient if there is no application code calling `Sidekiq.redis`, but it seems a good way to surface a very common configuration error a large fraction of the time.

**Comment by MrJoy** [comment]
@mperham I think this shows pretty definitively that this issue was caused by user error.  Apologies for wasting your time.

<img width="322" alt="image" src="https://github.com/sidekiq/sidekiq/assets/113625/b8ea6f90-038a-4562-b638-a968ef453f37">

**Comment by stanhu** [comment]
I submitted #6158 to add an extra 5 connections to the capsule pool by default.

**Comment by mperham** [comment]
If you are seeing issues with the concurrent limiter, we did just find a Redis < 7.0.8 bug in #6159 which can cause ReadTimeoutErrors. The solution is to upgrade Redis to the latest.

**Comment by MrJoy** [comment]
It is worth noting that apparently the latest-and-greatest Redis available via ElastiCache is... 7.0.7.  There's a 7.1 series but that's Amazon's proprietary patches on top of 7.0 and it's not clear to me if the underlying Redis is newer or not.  Hrm...
